brains/andalite.py

```diff
@@ -0,0 +1,107 @@
+import ast
+import asyncio
+import concurrent
+import json
+import pprint
+import queue
+import threading
+import websockets
+
+from sapiens import Sapiens
+
+class Andalite(Sapiens):
+    name = 'andalite'
+    lastSent = {}
+
+    def __init__(self, game, args=None):
+        Sapiens.__init__(self, game, args)
+        self.telepathy = Telepathy()
+
+    def Step(self):
+        self.sendMemory()
+        self.receiveBytes()
+        return Sapiens.Step(self)
+
+    def Event(self, evt):
+        Sapiens.Event(self, evt)
+
+    def Path(self):
+        return Sapiens.Path(self)
+
+    def Victory(self):
+        return Sapiens.Victory(self)
+
+    def sendMemory(self):
+        self.sendNewBytes(0xcc24, 11) # menu data
+
+    def sendNewBytes(self, start, lengthBytes):
+        data = {'offset': start, 'data': self.game.PeekMemoryRegion(start, lengthBytes)} # menu data
+        lastSentData = self.lastSent.get(start)
+        if lastSentData != data:
+            self.telepathy.outboundMemoryReads.put_nowait(str(data))
+            self.lastSent[start] = data
+
+    def receiveBytes(self):
+        try:
+            data = self.telepathy.inboundMemoryWrites.get_nowait()
+            start = data.get('offset')
+            lastData = self.lastSent.get(start)
+
+            if lastData == None or start != None and lastData.get('data') != data.get('data'):
+                print("writing data")
+                self.game.PokeMemoryRegion(data['offset'], data['data'])
+                self.lastSent['offset'] = data
+        except queue.Empty:
+            pass
+
+class Telepathy:
+    inboundMemoryWrites = queue.Queue()
+    outboundMemoryReads = queue.Queue()
+
+    def __init__(self):
+        thread = threading.Thread(target=self.threadMain, args=())
+        thread.daemon = True
+        thread.start()
+        #asyncio.get_event_loop().run_forever()
+    def threadMain(self):
+        loop = asyncio.new_event_loop()
+        #loop.run_in_executor(executor, self.handleSocket)
+        loop.create_task(self.handleSocket())
+        loop.run_forever()
+
+
+    async def handleSocket(self):
+        async with websockets.connect('ws://localhost:8765') as websocket:
+            print("connected to websocket")
+            consumerTask = asyncio.ensure_future(self.messageConsumer(websocket))
+            producerTask = asyncio.ensure_future(self.messageProducer(websocket))
+            done, pending = await asyncio.wait(
+                    [consumerTask, producerTask],
+                    return_when=asyncio.FIRST_COMPLETED)
+            for task in pending:
+                task.cancel()
+
+    async def messageConsumer(self, websocket):
+        while True:
+            message = await websocket.recv()
+            data = ast.literal_eval(message)
```

Comment:
this'll do for now, but we may want to think about a different wire format than hoping `eval(str(x))` remains an identity function for all the types.

Reply:
agreed. this is also pretty inefficient tbh

Reply:
created #5